メモアプリの実装

guchi5 opened the pull request:
jsonファイルにメモを保存する形式でメモアプリを実装しました。

guchi5 committed:
メモアプリの実装
- jsonファイルに保存する形式でメモアプリを実装
- README.mdの追記

guchi5 committed:
README.mdを更新

guchi5 committed:
.gitignoreの更新

guchi5 committed:
Revert ".gitignoreの更新"

This reverts commit 8ee965d0b11770ed0800026524a39e99ce0fe3ba.

guchi5 committed:
erbのpostの不具合を解消

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,102 @@
+end
+
+get '/' do
+  @memos = File.open('memo.json', 'r') do |file|
+    data = file.read
+    if data == ''
+      {}
+    else
+      JSON.parse(data)
+    end
+  end
```
このあたり共通部分はメソッド化してみましょう。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,102 @@
+  memos = File.open('memo.json', 'r') do |file|
+    data = file.read
+    if data == ''
+      JSON.parse('{}')
```
わざわざ `JSON.parse` しなくても `{}` で空のHash返せばいいのではと思いました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,102 @@
+  end
+
+  id = if !memos.empty?
+         memos.keys.last.to_i + 1
```
ここではこれでもよいですが、一般的にはこの方法だと問題があります。

- 同じIDが複数回発行されることがある(一旦最大のidのものを削除して再度新規作成した場合など)ので、たとえばURLでブックマークしていると、知らない間に内容が差しかわっている可能性がでてしまう
- 一般的にWebサーバーは複数台で運用されるので、同時にリクエストがくると同じidで複数のメモが作られてしまう。

ではどう解決するかというと、よくとられる方法は以下のような感じです。

- idを一意に発行できる仕組みを作る、使う。一度発行したidは二度と発行しない。
- uuidなどそもそも(現実的には)一意性を担保できるものをidとして使う

**修正は不要**ですが、頭の片隅にでもいれておいてください。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,102 @@
+  end
+end
+
+get '/' do
```
プラクティスの要件に

> RESTに倣ったURIになっていること

というのがあります。ここに限らず、RESTになっていないところが多いので、Path設計は全体的に見直してみてください。

yoshitsugu commented on memo.json:
```diff
@@ -0,0 +1 @@
+{"0":{"title":"ddd","text":"dssdf"}}
```
リポジトリに含めるのは空のデータにするか、データ用のファイルはリポジトリに含めないようにしましょう。

yoshitsugu commented on public/edit/css/reset.css:
reset.cssはなぜ複数含まれているのでしょう？ファイルとしては1つにして共通して参照するのではダメでしょうか？

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,102 @@
+end
+
+post '/save' do
+  @title = h(params[:title])
+  @text = h(params[:text])
```
HTMLエスケープ処理は基本、view側で行います。
なぜかというと、元のテキストが失われてしまうので、記録的な意味でそのままとっておきたいのが一点、
また、例えば、CSVなどHTML以外のフォーマットに出力したい、という要望が今後追加されたとすると、HTMLエスケープ処理は必要ないですよね。ということで、「HTMLエスケープしないといけない」というのはview側の都合なんですよね。なので、データ自体は改変せずに、view側で都度エスケープして表示する、というのが自然かなと思います！
SinatraのFAQにも[こんな感じのサンプルコード](https://sinatrarb.com/faq.html#escape_html)があります

yoshitsugu reviewed (commented):
コメントしました。

guchi5 committed:
コードの修正
- parse_jsonの追加
- RESTに倣ったURIに修正
- memo.jsonを空に修正
- reset.css呼び出し元を統一
- htmlエスケープ処理をview側で行うよう修正

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,101 @@
+    Rack::Utils.escape_html(text)
+  end
+
+  def parse_json
```
この後のプラクティスで保存先をDBにしてもらいますが、アプリケーションにとって、裏側がJSONであることは本質的な情報ではありません。
`fetch_memos` など、このメソッドが存在する「目的」を名前にするようにしましょう。「何をしているか」は隠蔽するほうがメソッドの役割にも適っています。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,101 @@
+end
+
+get '/memo/edit/:id' do |id|
+  memos = File.open('memo.json', 'r') do |file|
+    data = file.read
+    JSON.parse(data)
+  end
```
この辺もメソッドを共通利用できそうですね。
一つ指摘をうけたときは同じような箇所がないか、ご自身でも確認できるとよいですね。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,101 @@
+    JSON.parse(data)
+  end
+  memos.delete(id)
+  File.open('memo.json', 'w') do |file|
+    JSON.dump(memos, file)
+  end
```
書き込み側もメソッド化できそうではないですか？

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,101 @@
+  end
+end
+
+get '/memo' do
```
プラクティスの要件に

> RESTに倣ったURIになっていること

というのがあります。リソース一覧の取得をするときのRESTのPathはどうだったでしょうか？
他全般的にあまりRESTに倣ってないように見えますので、見直してみてください。

yoshitsugu replied:
まだRESTに倣ってないようです。
`/(リソース名の複数形)` もしくは`/(リソース名の複数形)/(リソースID)` が基本形だったことを思い出してください。

yoshitsugu reviewed (commented):
コメントしました。

guchi5 committed:
RESTに倣ったURIに変更

guchi5 committed:
プログラムの修正
- parse_jsonをread_memosに変更
- save_memosを追加

yoshitsugu commented on .gitignore:
```diff
@@ -0,0 +1 @@
+Gemfile.lock
```
通常 Gemfile.lockもリポジトリに含めます。なぜかというと、開発環境で試して、本番環境にデプロイするとき、gemのバージョンがあがってしまい、互換性が崩れて動かない、ということがあるからです。
Gemfile.lockを含めることで、インストールするgemのバージョンを固定できます。

guchi5 replied:
`.gitignore`から削除しました！

yoshitsugu commented on Gemfile:
```diff
@@ -0,0 +1,13 @@
+
+group :development do
+  gem 'erb_lint', require: false
+  gem 'puma'
```
pumaやsinatraは `development` ではなく、実際に本番環境でも必要なgemだと思うのですが、どうでしょうか。

guchi5 replied:
`production`グループに移行しました

yoshitsugu replied:
これだと `production`  でのみ必要なように見えます。
全ての環境で必要なものはgroup外に置くとよいです。

yoshitsugu commented on Gemfile:
```diff
@@ -0,0 +1,13 @@
+  gem 'erb_lint', require: false
+  gem 'puma'
+  gem 'rackup'
+  gem 'redcarpet'
```
redcarpetはなぜ必要なんでしょう？

guchi5 replied:
必要なさそう削除しました！

yoshitsugu reviewed (commented):
コメントしました。

guchi5 committed:
プログラムの修正
- RESTに倣ったURIに修正
- Gemfile.lockの追加
- Gemfileの修正

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,89 @@
+  erb :new_memo
+end
+
+post '/memos/new' do
```
リソース追加のPathがRESTに沿ってないですね。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,89 @@
+  erb :edit_memo
+end
+
+patch '/memos/:id/edit' do |id|
```
更新も違いそうです。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,89 @@
+  erb :show_memo
+end
+
+delete '/memos/:id/delete' do |id|
```
削除も違います。

yoshitsugu reviewed (commented):
コメントしました。
再度RESTの復習をおねがいします。

guchi5 committed:
プログラムの修正
- Gemfileの修正
- RESTに倣ったURIに修正

yoshitsugu commented on public/css/edit/style.css:
```diff
@@ -0,0 +1,61 @@
+    background-color: #d7d5d5;
+    border-radius: 10px;
+    border-style: none;
+}
```
自動生成されるファイルを除き、ファイル末尾に改行がないことをあらわすマークがでないようにおねがいします。
参考: https://bootcamp.fjord.jp/questions/1405

guchi5 replied:
vscodeの設定で末尾改行を追加するようにしました！

yoshitsugu reviewed (commented):
一点だけコメントしました。

guchi5 committed:
行末改行を追加

guchi5 committed:
.erblin.ymlの行末改行

guchi5 committed:
行末改行を修正

yoshitsugu reviewed (approved):
OKです！
